# **Data Collection Notebook**

## Objectives

* Fetch data from Kaggle and save it as raw data.
* Inspect the data and save it under outputs/datasets/collection

## Inputs

* Kaggle JSON file and the authentication token.

## Outputs

Generate Datasets:
* outputs/datasets/collection/HousePrice.csv
* outputs/datasets/collection/InheritedHouses.csv


## Additional Comments

In the workplace, projects are done using [Kaggle data](https://www.kaggle.com/datasets/codeinstitute/housing-prices-data): 
1. The first dataset is used to build our machine learning model(s). 
2. The second represent the 4 houses which the Total sale price of the inherited houses should be predicted.



---

# Install python packages in the notebooks

In [1]:
%pip install -r /workspace/Prediction-heritage-housing/requirements.txt

  Using cached numpy-1.18.5-cp38-cp38-manylinux1_x86_64.whl (20.6 MB)
  Using cached pandas-1.4.2-cp38-cp38-manylinux_2_17_x86_64.manylinux2014_x86_64.whl (11.7 MB)
  Using cached matplotlib-3.3.1-cp38-cp38-manylinux1_x86_64.whl (11.6 MB)
  Using cached seaborn-0.11.0-py3-none-any.whl (283 kB)
  Using cached ydata_profiling-4.4.0-py2.py3-none-any.whl (356 kB)
  Using cached plotly-4.12.0-py2.py3-none-any.whl (13.1 MB)
  Using cached ppscore-1.2.0.tar.gz (47 kB)
  Using cached streamlit-0.85.0-py2.py3-none-any.whl (7.9 MB)
  Using cached feature_engine-1.0.2-py2.py3-none-any.whl (152 kB)
  Using cached imbalanced_learn-0.8.0-py3-none-any.whl (206 kB)
  Using cached scikit_learn-0.24.2-cp38-cp38-manylinux2010_x86_64.whl (24.9 MB)
  Using cached xgboost-1.2.1-py3-none-manylinux2010_x86_64.whl (148.9 MB)
  Using cached yellowbrick-1.3-py3-none-any.whl (271 kB)
  Using cached Jinja2-3.1.1-py3-none-any.whl (132 kB)
  Using cached MarkupSafe-2.0.1-cp38-cp38-manylinux_2_5_x86_64.manylinux1_x86

# Change working directory

We need to change the working directory from its current folder to its parent folder
* We access the current directory with os.getcwd()

In [2]:
import os
current_dir = os.getcwd()
current_dir

'/workspace/Prediction-heritage-housing/jupyter_notebooks'

We want to make the parent of the current directory the new current directory
* os.path.dirname() gets the parent directory
* os.chir() defines the new current directory

In [3]:
os.chdir(os.path.dirname(current_dir))
print("You set a new current directory")

You set a new current directory


Confirm the new current directory

In [4]:
current_dir = os.getcwd()
current_dir

'/workspace/Prediction-heritage-housing'

# Fetch raw data from Kaggle

First we need to install kaggle to access the kaggle API and to fetch the raw data.

In [5]:
! pip install kaggle==1.5.12

     |████████████████████████████████| 58 kB 4.8 MB/s eta 0:00:011
     |████████████████████████████████| 78 kB 7.5 MB/s  eta 0:00:01
Using legacy 'setup.py install' for kaggle, since package 'wheel' is not installed.
    Running setup.py install for kaggle ... done
You should consider upgrading via the '/workspace/.pyenv_mirror/fakeroot/versions/3.8.12/bin/python3.8 -m pip install --upgrade pip' command.


Make the kaggle authentication token available for the session.

In [6]:
import os
os.environ['KAGGLE_CONFIG_DIR'] = os.getcwd()
! chmod 600 kaggle.json

* Define path to the Kaggle dataset we want to download
* Indicate the destination folder for the downloaded data
* Download the data

In [7]:
KaggleDatasetPath = "codeinstitute/housing-prices-data"
DestinationFolder = "inputs/datasets/raw"   
! kaggle datasets download -d {KaggleDatasetPath} -p {DestinationFolder}

  0%|                                               | 0.00/49.6k [00:00<?, ?B/s]
100%|██████████████████████████████████████| 49.6k/49.6k [00:00<00:00, 2.82MB/s]


* Unzip the downloaded folder
* Remove the zipped folder
* Remove the kaggle JSON file

In [8]:
! unzip {DestinationFolder}/*.zip -d {DestinationFolder} \
  && rm {DestinationFolder}/*.zip \
  && rm kaggle.json

Archive:  inputs/datasets/raw/housing-prices-data.zip
  inflating: inputs/datasets/raw/house-metadata.txt  
  inflating: inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/house_prices_records.csv  
  inflating: inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/inherited_houses.csv  


---

# Load and Inspect Kaggle data

* Import the pandas library
* Load the datasets as a pandas DataFrame and assign it to df for each datasets

## House Price Records DataFrame

In [10]:
import pandas as pd
df_houseprices = pd.read_csv(f"inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/house_prices_records.csv")
df_houseprices.head()

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       856     854.0           3.0           No         706          GLQ   
1      1262       0.0           3.0           Gd         978          ALQ   
2       920     866.0           3.0           Mn         486          GLQ   
3       961       NaN           NaN           No         216          ALQ   
4      1145       NaN           4.0           Av         655          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  ...  LotFrontage  \
0        150            0.0         548          RFn  ...         65.0   
1        284            NaN         460          RFn  ...         80.0   
2        434            0.0         608          RFn  ...         68.0   
3        540            NaN         642          Unf  ...         60.0   
4        490            0.0         836          RFn  ...         84.0   

   MasVnrArea OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  WoodDeckSF  \
0       196.0          61            5            7          856         0.0   
1         0.0           0            8            6         1262         NaN   
2       162.0          42            5            7          920         NaN   
3         0.0          35            5            7          756         NaN   
4       350.0          84            5            8         1145         NaN   

   YearBuilt  YearRemodAdd  SalePrice  
0       2003          2003     208500  
1       1976          1976     181500  
2       2001          2002     223500  
3       1915          1970     140000  
4       2000          2000     250000  

[5 rows x 24 columns]

In [11]:
df_houseprices.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1460 entries, 0 to 1459
Data columns (total 24 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1stFlrSF       1460 non-null   int64  
 1   2ndFlrSF       1374 non-null   float64
 2   BedroomAbvGr   1361 non-null   float64
 3   BsmtExposure   1460 non-null   object 
 4   BsmtFinSF1     1460 non-null   int64  
 5   BsmtFinType1   1346 non-null   object 
 6   BsmtUnfSF      1460 non-null   int64  
 7   EnclosedPorch  136 non-null    float64
 8   GarageArea     1460 non-null   int64  
 9   GarageFinish   1298 non-null   object 
 10  GarageYrBlt    1379 non-null   float64
 11  GrLivArea      1460 non-null   int64  
 12  KitchenQual    1460 non-null   object 
 13  LotArea        1460 non-null   int64  
 14  LotFrontage    1201 non-null   float64
 15  MasVnrArea     1452 non-null   float64
 16  OpenPorchSF    1460 non-null   int64  
 17  OverallCond    1460 non-null   int64  
 18  OverallQ

In [15]:
df_houseprices.drop(labels=['EnclosedPorch', 'WoodDeckSF'], axis=1)

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0          856     854.0           3.0           No         706          GLQ   
1         1262       0.0           3.0           Gd         978          ALQ   
2          920     866.0           3.0           Mn         486          GLQ   
3          961       NaN           NaN           No         216          ALQ   
4         1145       NaN           4.0           Av         655          GLQ   
...        ...       ...           ...          ...         ...          ...   
1455       953     694.0           3.0           No           0          Unf   
1456      2073       0.0           NaN           No         790          ALQ   
1457      1188    1152.0           4.0           No         275          GLQ   
1458      1078       0.0           2.0           Mn          49          NaN   
1459      1256       0.0           3.0           No         830          BLQ   

      BsmtUnfSF  GarageArea GarageFinish  GarageYrBlt  ...  LotArea  \
0           150         548          RFn       2003.0  ...     8450   
1           284         460          RFn       1976.0  ...     9600   
2           434         608          RFn       2001.0  ...    11250   
3           540         642          Unf       1998.0  ...     9550   
4           490         836          RFn       2000.0  ...    14260   
...         ...         ...          ...          ...  ...      ...   
1455        953         460          RFn       1999.0  ...     7917   
1456        589         500          Unf       1978.0  ...    13175   
1457        877         252          RFn       1941.0  ...     9042   
1458          0         240          Unf       1950.0  ...     9717   
1459        136         276          Fin       1965.0  ...     9937   

     LotFrontage  MasVnrArea  OpenPorchSF  OverallCond  OverallQual  \
0           65.0       196.0           61            5            7   
1           80.0         0.0            0            8            6   
2           68.0       162.0           42            5            7   
3           60.0         0.0           35            5            7   
4           84.0       350.0           84            5            8   
...          ...         ...          ...          ...          ...   
1455        62.0         0.0           40            5            6   
1456        85.0       119.0            0            6            6   
1457        66.0         0.0           60            9            7   
1458        68.0         0.0            0            6            5   
1459        75.0         0.0           68            6            5   

      TotalBsmtSF  YearBuilt  YearRemodAdd  SalePrice  
0             856       2003          2003     208500  
1            1262       1976          1976     181500  
2             920       2001          2002     223500  
3             756       1915          1970     140000  
4            1145       2000          2000     250000  
...           ...        ...           ...        ...  
1455          953       1999          2000     175000  
1456         1542       1978          1988     210000  
1457         1152       1941          2006     266500  
1458         1078       1950          1996     142125  
1459         1256       1965          1965     147500  

[1460 rows x 22 columns]

DATAFRAME SUMMARY: 
- We had a dataframe of 1460 rows and 24 colums.
- As EnclosedPorch and WoodDeckSF were missing more than half the data, the colums were drop in the first inspection.
- There is 3 different data types : int64, float64 and object.


## Inherited Houses DataFrame

In [16]:
df_inherited = pd.read_csv(f"inputs/datasets/raw/house-price-20211124T154130Z-001/house-price/inherited_houses.csv")
df_inherited

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       896         0             2           No       468.0          Rec   
1      1329         0             3           No       923.0          ALQ   
2       928       701             3           No       791.0          GLQ   
3       926       678             3           No       602.0          GLQ   

   BsmtUnfSF  EnclosedPorch  GarageArea GarageFinish  ...  LotArea  \
0      270.0              0       730.0          Unf  ...    11622   
1      406.0              0       312.0          Unf  ...    14267   
2      137.0              0       482.0          Fin  ...    13830   
3      324.0              0       470.0          Fin  ...     9978   

   LotFrontage MasVnrArea  OpenPorchSF  OverallCond  OverallQual  TotalBsmtSF  \
0         80.0        0.0            0            6            5        882.0   
1         81.0      108.0           36            6            6       1329.0   
2         74.0        0.0           34            5            5        928.0   
3         78.0       20.0           36            6            6        926.0   

   WoodDeckSF  YearBuilt  YearRemodAdd  
0         140       1961          1961  
1         393       1958          1958  
2         212       1997          1998  
3         360       1998          1998  

[4 rows x 23 columns]

In [14]:
df_inherited.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4 entries, 0 to 3
Data columns (total 23 columns):
 #   Column         Non-Null Count  Dtype  
---  ------         --------------  -----  
 0   1stFlrSF       4 non-null      int64  
 1   2ndFlrSF       4 non-null      int64  
 2   BedroomAbvGr   4 non-null      int64  
 3   BsmtExposure   4 non-null      object 
 4   BsmtFinSF1     4 non-null      float64
 5   BsmtFinType1   4 non-null      object 
 6   BsmtUnfSF      4 non-null      float64
 7   EnclosedPorch  4 non-null      int64  
 8   GarageArea     4 non-null      float64
 9   GarageFinish   4 non-null      object 
 10  GarageYrBlt    4 non-null      float64
 11  GrLivArea      4 non-null      int64  
 12  KitchenQual    4 non-null      object 
 13  LotArea        4 non-null      int64  
 14  LotFrontage    4 non-null      float64
 15  MasVnrArea     4 non-null      float64
 16  OpenPorchSF    4 non-null      int64  
 17  OverallCond    4 non-null      int64  
 18  OverallQual   

In [17]:
df_inherited.drop(labels=['EnclosedPorch', 'WoodDeckSF'], axis=1)

1stFlrSF  2ndFlrSF  BedroomAbvGr BsmtExposure  BsmtFinSF1 BsmtFinType1  \
0       896         0             2           No       468.0          Rec   
1      1329         0             3           No       923.0          ALQ   
2       928       701             3           No       791.0          GLQ   
3       926       678             3           No       602.0          GLQ   

   BsmtUnfSF  GarageArea GarageFinish  GarageYrBlt  ...  KitchenQual LotArea  \
0      270.0       730.0          Unf       1961.0  ...           TA   11622   
1      406.0       312.0          Unf       1958.0  ...           Gd   14267   
2      137.0       482.0          Fin       1997.0  ...           TA   13830   
3      324.0       470.0          Fin       1998.0  ...           Gd    9978   

   LotFrontage  MasVnrArea  OpenPorchSF  OverallCond  OverallQual  \
0         80.0         0.0            0            6            5   
1         81.0       108.0           36            6            6   
2         74.0         0.0           34            5            5   
3         78.0        20.0           36            6            6   

   TotalBsmtSF  YearBuilt  YearRemodAdd  
0        882.0       1961          1961  
1       1329.0       1958          1958  
2        928.0       1997          1998  
3        926.0       1998          1998  

[4 rows x 21 columns]

DATAFRAME SUMMARY: 
- We had a dataframe of 4 rows and 23 colums representing the 4 houses.
- As EnclosedPorch and WoodDeckSF were drop before, we droped them again. 
- There is 3 different data types : int64, float64 and object.


---

# Push files to Repo

In this notebook, we have collected kaggle data and inspected the columns in the dataset.

* From the quick inspection of the data, we can already did a first cleaning.
* We also see that the number of colums are different between the two dataframes.

We push the datasets to Repo.

In [18]:
import os
try:
  os.makedirs(name='outputs/datasets/collection') # create outputs/datasets/collection folder
except Exception as e:
  print(e)

df_houseprices.to_csv(f"outputs/datasets/collection/HousePrice.csv", index=False)
df_inherited.to_csv(f"outputs/datasets/collection/InheritedHouses.csv", index=False)